File: coding_int/chapter_2/2_5_Sum_Lists.cpp
// need to handle teens properly

#include<iostream>
#include<vector>
#include<string>
#include<sstream>

class node {
    public:

    node* next;
    int data_value;
};

class llist {

    private:

    node* head;
    node* tail;

    public:

    llist() {
        head=NULL;
        tail=NULL;
    }

    node* getHead(){
        return head;
    }

    void addValue(int adder) {
        
        node *temp = new node;
        temp->data_value = adder;
        temp->next=NULL;

        if (head==NULL) {
            head = temp;
            tail = temp; 
        } else {
            tail->next = temp;
            tail = temp;
        }
    }

    void printList() {
        
        node *temp = new node;
        temp = head;

        while(temp) {
            std::cout << temp->data_value << "\t";
            temp = temp->next;
        }
        std::cout << std::endl << "---end---" << std::endl;
    }
};

void sum_vecs(llist *one, llist *two, llist *holder) {

    node *temp_1 = new node;
    node *temp_2 = new node;

    temp_1 = one->getHead();
    temp_2 = two->getHead();

    std::string holder1 = "";
    std::string holder2 = "";
    std::string digit = "";

    while(temp_1) {

        digit = std::to_string(temp_1->data_value);
        holder1 = digit + holder1;
        temp_1 = temp_1->next;
    }

    while(temp_2) {
        digit = std::to_string(temp_2->data_value);
        holder2 = digit + holder2;
        temp_2 = temp_2->next;
    }
        
    int result = std::stoi(holder1) + std::stoi(holder2);
    
    std::string result_str = std::to_string(result);
    std::string reverse_str;
    
    //std::cout << "before reorder: " << result_str << std::endl;

    for (auto item : result_str) {
        reverse_str = item + reverse_str;
    }
    
    for (char& c : reverse_str) {
        int out = c - '0';
        holder->addValue(out);
    }

    //std::cout << reverse_str << std::endl;
}

int main() {

    llist *vector_1 = new llist;
    llist *vector_2 = new llist;

    vector_1->addValue(4);
    vector_1->addValue(5);
    vector_1->addValue(6);

    vector_2->addValue(3);
    vector_2->addValue(2);
    vector_2->addValue(7);

    llist *answer = new llist;
    sum_vecs(vector_1, vector_2, answer);
    answer->printList();



}

File: coding_int/chapter_2/2_5_Sum_Lists.py
class node:
    def __init__(self, value = None):
        self.next = None
        self.value = value

class llist:
    def __init__(self, head = None):
        self.head = head
        self.tail = None

    def __iter__(self):
        current = self.head
        while current is not None:
            yield current
            current = current.next

    def __str__(self):    
        for item in self:
            if item is not None:
                print("{0} /t".format(item.value))

        return "---END---"

    def addValue(self, value):
        
        new_node = node(value)
        if (self.head is None):
            self.head = new_node
            self.tail = new_node

        else:
            self.tail.next = new_node
            self.tail = new_node


def sum_vecs(list1, list2):

    total = llist()

    for i1, i2 in zip(list1, list2):
        v1 = i1.value
        v2 = i2.value

        sum_val = v1+v2
        total.addValue(sum_val)
    
    return total


if __name__ == "__main__":

    vec_1 = llist()
    vec_1.addValue(3)
    vec_1.addValue(5)
    vec_1.addValue(4)

    vec_2 = llist()
    vec_2.addValue(6)
    vec_2.addValue(4)
    vec_2.addValue(3)

    sum_vec = sum_vecs(vec_1, vec_2)
    print(sum_vec)